### ISSUE TITLE
Potential problem in calculateConfusionMatrix()

### ISSUE THREAD

**Issue by dkyleward** [mlr-org/mlr#2030]
First I'm going to apologize, as I'm not sure how to create a minimum reproducible example for this issue. I absolutely own that it may be something that I've done wrong in my (lengthy) analysis, but maybe this can just be on your radar to look out for?

The general problem is that `calculateConfusionMatrix()` is getting confused over factor levels in the result.

When I run a basic table on my data, I get the result below. The true values run down the columns and note that in my 100-sample test set, I did not have any observations of class 1 or 4 (they do exist in my training data).

```r
table(xgb_result$data$truth, xgb_result$data$response)
```
![screen shot 2017-10-09 at 4 22 59 pm](https://user-images.githubusercontent.com/10221750/31357004-23664d0c-ad0e-11e7-899f-b847a1d7b15d.png)

When I run `calculateConfusionMatrix()`, I get the result below. Note that instead of having classes 1 and 4 missing, it is 9 and 10. The rest of the table is also messed up. (Note that in the above table, there were 11 correct classifications of class 5. Now that 11 is associated with a true value of 3.)
```r
calculateConfusionMatrix(xgb_result)
```
![screen shot 2017-10-09 at 4 24 17 pm](https://user-images.githubusercontent.com/10221750/31357041-525989e4-ad0e-11e7-9603-07b29f42378c.png)

To make sure I wasn't crazy, I wrote the table out to CSV and simply looked at the truth/response columns. The estimated model is actually doing a good job and gets the 5th class right 11 times (as shown by the `table()` method).

It's a sad consolation prize compared to a reproducible example, but here is my session info as requested:
```
R version 3.4.0 (2017-04-21)
Platform: x86_64-apple-darwin16.5.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] bindrcpp_0.2       censusr_0.0.3      httr_1.2.1        
 [4] leaflet_1.1.0      RColorBrewer_1.1-2 sf_0.5-1          
 [7] parallelMap_1.3    plotly_4.7.0       kableExtra_0.4.0  
[10] knitr_1.17         mlr_2.12           ParamHelpers_1.10 
[13] MASS_7.3-47        dplyr_0.7.3        purrr_0.2.3       
[16] readr_1.1.0        tidyr_0.7.1        tibble_1.3.4      
[19] ggplot2_2.2.1      tidyverse_1.1.1    pixiedust_0.8.1   
[22] h2o_3.10.5.3      

loaded via a namespace (and not attached):
 [1] nlme_3.1-131        bitops_1.0-6        lazyWeave_3.0.1    
 [4] lubridate_1.6.0     rprojroot_1.2       tools_3.4.0        
 [7] backports_1.1.0     R6_2.2.2            rpart_4.1-11       
[10] Hmisc_4.0-3         DBI_0.6-1           lazyeval_0.2.0     
[13] colorspace_1.3-2    nnet_7.3-12         tidyselect_0.2.0   
[16] gridExtra_2.2.1     mnormt_1.5-5        compiler_3.4.0     
[19] rvest_0.3.2         htmlTable_1.9       xml2_1.1.1         
[22] scales_0.4.1        checkmate_1.8.3     psych_1.7.3.21     
[25] stringr_1.2.0       digest_0.6.12       foreign_0.8-67     
[28] rmarkdown_1.6       base64enc_0.1-3     pkgconfig_2.0.1    
[31] htmltools_0.3.6     htmlwidgets_0.8     rlang_0.1.2        
[34] readxl_1.0.0        BBmisc_1.11         shiny_1.0.3        
[37] bindr_0.1           jsonlite_1.4        crosstalk_1.0.0    
[40] acepack_1.4.1       RCurl_1.95-4.8      magrittr_1.5       
[43] Formula_1.2-1       Matrix_1.2-9        Rcpp_0.12.12       
[46] munsell_0.4.3       stringi_1.1.5       yaml_2.1.14        
[49] plyr_1.8.4          grid_3.4.0          parallel_3.4.0     
[52] forcats_0.2.0       udunits2_0.13       lattice_0.20-35    
[55] haven_1.0.0         splines_3.4.0       hms_0.3            
[58] xgboost_0.6-4       reshape2_1.4.2      glue_1.1.1         
[61] evaluate_0.10.1     latticeExtra_0.6-28 data.table_1.10.4  
[64] modelr_0.1.0        httpuv_1.3.3        cellranger_1.1.0   
[67] gtable_0.2.0        assertthat_0.2.0    mime_0.5           
[70] xtable_1.8-2        broom_0.4.2         survival_2.41-3    
[73] viridisLite_0.2.0   units_0.4-6         cluster_2.0.6 
```

**Comment by larskotthoff** [comment]
Thanks for the report. Can you post the code that you ran to get `xgb_result` please (including any data)?

**Comment by dkyleward** [comment]
Here is my code. Unfortunately, my data is privacy-sensitive survey material. There is substantial pre-processing that happens before this code snippet that creates `inc_data`, but I think this should be enough to get the idea of what I did. Thanks!

```r
# Set aside the records without income information.
# These are the records that need imputation. The records with income info
# will become train and test data sets.
to_impute <- inc_data %>%
  filter(is.na(INCOME_RANGE)) %>%
  select(-HHID) %>%
  mutate(INCOME_RANGE = factor("1", levels = as.character(1:10)))
est_data <- inc_data %>%
  filter(!is.na(INCOME_RANGE))

# create train and test data frames
# train_with_hhid <- est_data[1:(nrow(est_data) - 100), ]
# test_with_hhid <- est_data[(nrow(est_data) - 100):nrow(est_data), ]
test_with_hhid <- est_data %>%sample_n(100, weight = as.numeric(INCOME_RANGE))
train_with_hhid <- est_data[!est_data$HHID %in% test_with_hhid$HHID, ]

# Remove the HHID so it won't be considered a feature
train <- train_with_hhid %>%
  select(-HHID)
test <- test_with_hhid %>%
  select(-HHID)

# Create tasks
trainTask <- makeClassifTask(data = train, target = "INCOME_RANGE")
testTask <- makeClassifTask(data = test, target = "INCOME_RANGE")
imputeTask <- makeClassifTask(data = to_impute, target = "INCOME_RANGE")

# Create learner
xgb_learner <- makeLearner(
  "classif.xgboost",
  predict.type = "prob",
  par.vals = list(
    # mlr folks noted bug when objective is included (issue created). remove.
    # objective = "multi:softmax"
  )
)

# Make a hyper-parameter set for tuning
xgb_params <- makeParamSet(
  # The number of trees in the model (each one built sequentially)
  makeIntegerParam("nrounds", lower = 100, upper = 250),
  # number of splits in each tree
  makeIntegerParam("max_depth", lower = 1, upper = 3),
  # "shrinkage" - prevents overfitting
  makeNumericParam("eta", lower = .1, upper = .5),
  # L2 regularization - prevents overfitting
  makeNumericParam("lambda", lower = -1, upper = 0, trafo = function(x) 10^x)
)

# Create a control for searching the hyper-parameter space
control <- makeTuneControlRandom(maxit = 24)

# Create a resampling plan
resample_desc <- makeResampleDesc("CV", iters = 3)

# Tune the hyper-parameters
parallelStartMulticore(8)
tuned_params <- tuneParams(
  learner = xgb_learner,
  task = trainTask,
  resampling = resample_desc,
  par.set = xgb_params,
  control = control
)
parallelStop()

# Create a new learner using the tuned hyper-parameters
xgb_learner_tuned <- setHyperPars(
  learner = xgb_learner,
  par.vals = tuned_params$x
)

# Create model
xgb_model <- train(xgb_learner_tuned, task = trainTask)

**Comment by larskotthoff** [comment]
What's the structure of your data, in particular what are the classes? It looks like your target is always `factor("1")`.

**Comment by berndbischl** [comment]
@dkyleward in your last code there is not even a call into calculateConfusionMatrix? or the creation of a prediction object?

also some other options: 

a) pls check whether the error in in the output of calculateConfusionMatrix or the printer. simply unclass the result, that should show a bare list, with a matrix inside.

b) if you cannot find a simple example, with different data (did you try that at least?): why not run through calculateConfusionMatrix in single step mode? and tell us what goes wrong? the code code is very simple and an "endpoint" and does not call 1000 other functions recursively?

**Comment by dkyleward** [comment]
I had some more time today (and honestly I had to learn more about debugging in R - discovering the trace function was a godsend). I figured out where the problem occurs in the function. Also, while I can't give raw data, there's no problem giving you my resulting object from the xgb model. You can call the function on that to see exactly what I'm seeing.

[xgb_result.RDS.zip](https://github.com/mlr-org/mlr/files/1373373/xgb_result.RDS.zip)

When I trace through the function, everything looks right until the following line:
`dimnames(result) = list(true = c(cls, "-err.-"), predicted = c(cls, "-err.-"))`

Here is what `truth`, `resp`, and `result` look like right before that line. Note that the levels of `truth` and `resp` aren't identical.
<img width="570" alt="screen shot 2017-10-10 at 4 39 25 pm" src="https://user-images.githubusercontent.com/10221750/31409368-a82a6e8a-add9-11e7-8cfc-777645820053.png">

After running `dimnames`, it looks like this:
<img width="288" alt="screen shot 2017-10-10 at 4 41 47 pm" src="https://user-images.githubusercontent.com/10221750/31409444-ef82a2b6-add9-11e7-90fb-aa74ef207e6e.png">

It's the assertion of `cls` for both `true` and `predicted` that causes the problem. Instead, the following code works for me:

```r
rws <- rownames(result)
rws[length(rws)] <- "-err.-"
cls <- colnames(result)
cls[length(rws)] <- "-err.-"
dimnames(result) = list(true = rws, predicted = cls)
```

<img width="289" alt="screen shot 2017-10-10 at 4 57 07 pm" src="https://user-images.githubusercontent.com/10221750/31410185-168a664e-addc-11e7-92e4-180dd6b43b4f.png">

If you want, I can fork and pull request. It's a small enough change you may prefer to do it yourself. Let me know, and thanks for your patience.

**Comment by larskotthoff** [comment]
Should be fixed in #2033. It looks like you're passing the data you predict on as a separate data frame. You can avoid these kinds of problems if you use a single task for everything and subset.

**Comment by dkyleward** [comment]
I didn't know what you were talking about, but I saw it in the basic tutorial [here](https://mlr-org.github.io/mlr-tutorial/devel/html/). Nice feature, and thanks for the tip!
